[WebXR][OpenXR] Add basic frame submission to the render loop

svillar opened the pull request:
#### 6195d809d659cbd8731d80dfe378b0902af05db5
<pre>
[WebXR][OpenXR] Add basic frame submission to the render loop
<a href="https://bugs.webkit.org/show_bug.cgi?id=295507">https://bugs.webkit.org/show_bug.cgi?id=295507</a>

Reviewed by Carlos Garcia Campos.

So far the OpenXR render loop has been polling OpenXR events but nothing else. This adds
the machinery for submitting frames. This involves implementing both the
scheduleAnimationFrame() and the submitFrame() methods from the PlatformXRCoordinator.
The former stores the callback coming from WebXR in onFrameUpdate for the render loop
to call it, while the latter is the one actually submitting the frames to OpenXR runtime.

So far we&apos;re submitting mostly empty FrameData objects to WebXR but we&apos;ll eventually
fill the proper data in in follow up patches. OpenXR frame submission requires specifying
a blend mode though. That&apos;s why a blend mode selector code was added. It tries to select
the best blend mode for AR and VR modes based on the device supported blend modes.

* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::endSessionIfExists):
(WebKit::OpenXRCoordinator::scheduleAnimationFrame):
(WebKit::OpenXRCoordinator::submitFrameInternal):
(WebKit::OpenXRCoordinator::submitFrame):
(WebKit::OpenXRCoordinator::initializeDevice):
(WebKit::OpenXRCoordinator::initializeBlendModes):
(WebKit::OpenXRCoordinator::handleSessionStateChange):
(WebKit::OpenXRCoordinator::pollEvents):
(WebKit::OpenXRCoordinator::renderLoop):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: <a href="https://commits.webkit.org/297048@main">https://commits.webkit.org/297048@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2f20f052991a7187085b654a25e2254696160b0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110436 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30095 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20528 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116461 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60690 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30774 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38682 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83981 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113384 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24562 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99442 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64422 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23925 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60257 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93946 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17637 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119252 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37476 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27809 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92948 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37849 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95710 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92771 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23626 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37772 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15507 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33422 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37371 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42842 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37033 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40373 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38741 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -285,6 +347,36 @@ void OpenXRCoordinator::initializeDevice()
+    ASSERT(count);
+
+    Vector<XrEnvironmentBlendMode> blendModes;
+    blendModes.resize(count);
```
```suggestion
    Vector<XrEnvironmentBlendMode> blendModes(count);
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -354,22 +449,22 @@ void OpenXRCoordinator::handleSessionStateChange()
 }
 
-void OpenXRCoordinator::pollEvents(std::atomic<bool>& terminateRequested)
+void OpenXRCoordinator::pollEvents(bool& exitRenderLoop, Box<RenderState> active)
```
Could we make it return bool instead of passing reference to a bool?

svillar replied:
Yeah, but it looked weird to me to return a boolean in a method called pollEvents(). You'd always had to check the method to know what that boolean is

carlosgcampos replied:
We can return an enum like `enum class PollResult : bool { Continue, Stop }`

svillar replied:
fair enough, let's do that

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:
```diff
@@ -50,21 +50,35 @@ class OpenXRCoordinator final : public PlatformXRCoordinator {
+    void handleSessionStateChange(Box<RenderState>);
     void endSessionIfExists(std::optional<WebCore::PageIdentifier>);
-    void pollEvents(std::atomic<bool>& terminateRequested);
+    enum class PollResult : bool;
```
Why don't you just define it here?

svillar replied:
Well it's an implementation detail, I though it'd be a good idea to "hide" it in the header, but I can do it here if you prefer.

svillar committed:
[WebXR][OpenXR] Add basic frame submission to the render loop
https://bugs.webkit.org/show_bug.cgi?id=295507

Reviewed by Carlos Garcia Campos.

So far the OpenXR render loop has been polling OpenXR events but nothing else. This adds
the machinery for submitting frames. This involves implementing both the
scheduleAnimationFrame() and the submitFrame() methods from the PlatformXRCoordinator.
The former stores the callback coming from WebXR in onFrameUpdate for the render loop
to call it, while the latter is the one actually submitting the frames to OpenXR runtime.

So far we're submitting mostly empty FrameData objects to WebXR but we'll eventually
fill the proper data in in follow up patches. OpenXR frame submission requires specifying
a blend mode though. That's why a blend mode selector code was added. It tries to select
the best blend mode for AR and VR modes based on the device supported blend modes.

* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::endSessionIfExists):
(WebKit::OpenXRCoordinator::scheduleAnimationFrame):
(WebKit::OpenXRCoordinator::submitFrameInternal):
(WebKit::OpenXRCoordinator::submitFrame):
(WebKit::OpenXRCoordinator::initializeDevice):
(WebKit::OpenXRCoordinator::initializeBlendModes):
(WebKit::OpenXRCoordinator::handleSessionStateChange):
(WebKit::OpenXRCoordinator::pollEvents):
(WebKit::OpenXRCoordinator::renderLoop):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: https://commits.webkit.org/297048@main
